compiler/codegen/mli.cpp

```diff
@@ -0,0 +1,912 @@
+  return gen;
+}
+  
+std::string
+MLIContext::genServerDispatchSwitch(const std::vector<FnSymbol*>& fns) {
+  std::string gen;
+
+  gen += "int64_t chpl_mli_sdispatch";
+  gen += "(int64_t function)";
+  gen += scope_begin;
+  gen += this->genNewDecl("int", "err");
+  gen += "switch (function)";
+  gen += scope_begin;
+
+  for_vector(FnSymbol, fn, fns) {
+    if (not fn->hasFlag(FLAG_EXPORT) || fn->hasFlag(FLAG_GEN_MAIN_FUNC)) {
+      continue;
+    }
+
+    gen += "case ";
+    gen += this->genFuncNumericID(fn);
+    gen += ": ";
+
+    gen += scope_begin;
+    
+    if (this->injectDebugPrintlines) {
+      gen += this->genDebugPrintCall(fn, "[Server]");
+    }
+    
+    gen += "err = ";
+    gen += this->genServerWrapperCall(fn);
+    gen += scope_end;
+    gen += "break;\n";
+  }
+
+  gen += "default: return CHPL_MLI_ERROR_NOFUNC; break;\n";
+  gen += scope_end;
+  gen += "return err;\n";
+  gen += scope_end;
+
+  return gen;
+}
+
+//
+// TODO: This is unused right now (no struct type support yet).
+//
+bool MLIContext::structContainsOnlyPrimitiveScalars(Type *t) {
+  return false;
+}
+
+//
+// TODO: This filter will change as we support more and more type classes.
+//  - [ ] dtStringC
+//  - [ ] Chapel `string` type
+//  - [ ] ???
+//
+bool MLIContext::isSupportedType(Type* t) {
+  return (isPrimitiveScalar(t));
+}
+
+void MLIContext::verifyPrototype(FnSymbol* fn) {
+
+  if (fn->retType != dtVoid && not isSupportedType(fn->retType)) {
+    USR_FATAL("MLI does not support code generation for type", fn->retType);
+  }
+
+  // Loop through all formals and error gen if a type is unsupported.
+  for (int i = 1; i <= fn->numFormals(); i++) {
+    ArgSymbol* as = fn->getFormal(i);
+    if (isSupportedType(as->type)) { continue; }
+    USR_FATAL("MLI does not support code generation for type", as->type);
+  }
+
+  return;
+}
+
+Type* MLIContext::getTypeFromFormal(ArgSymbol* as) {
+  if (as == NULL) { return NULL; }
+  return as->type;
+}
+
+Type* MLIContext::getTypeFromFormal(FnSymbol* fn, int i) {
+  return getTypeFromFormal(fn->getFormal(i));
+}
+
+std::string MLIContext::genClientsideRPC(FnSymbol* fn) {
+  bool hasVoidReturnType = fn->retType == dtVoid;
+  bool hasFormals = fn->numFormals() != 0;
+  std::string gen;
+
+  // Declare the unique ID for this function.
+  gen += "int64_t id = ";
+  gen += this->genFuncNumericID(fn);
+  gen += ";\n";
+
+  // Declare a int64 for the server return value.
+  gen += this->genNewDecl("int64_t", "st");
+
+  // Declare a temporary for the return value, if necessary.
+  if (not hasVoidReturnType) {
+    gen += this->genNewDecl(fn->retType, "result");
+  }
+
+  if (this->injectDebugPrintlines) {
+    gen += this->genDebugPrintCall(fn, "[Client]");
+  }
+
+  // Push function to call.
+  gen += this->genSocketPushCall(client_main, "id");
+
+  // Pull server confirmation.
+  gen += this->genSocketPullCall(client_main, "st");
+
+  // TODO: Handle server errors.
+  gen += this->genTodo("Handle server errors.");
```

Comment:
I'm not sure I would actually emit this code.  Handling errors thrown by the server is a *long* ways down the road, if we even do it.  I'm more okay leaving the TODO itself in the chapel compiler.

Reply:
Fair. That was just for me when reading the generated code.